api/src/pcapi/core/offers/repository.py

```diff
@@ -550,6 +465,140 @@ def _filter_by_status(query: BaseQuery, status: str) -> BaseQuery:
+        if DisplayedStatus.PREBOOKED.value in statuses:
+            allowed_booking_status.add(educational_models.CollectiveBookingStatus.PENDING.value)
+
+        status_with_booking_filters.append(
+            and_(
+                offer_id_query.c.status.in_(allowed_booking_status),
+                educational_models.CollectiveOffer.status != offer_mixin.CollectiveOfferStatus.EXPIRED.name,
+            )
+        )
+    if DisplayedStatus.ENDED.value in statuses:
+        # Status ENDED == event is passed with a reservation not cancelled
+
+        status_with_booking_filters.append(
+            and_(
+                offer_id_query.c.status.in_(
+                    [
+                        educational_models.CollectiveBookingStatus.USED.value,
+                        educational_models.CollectiveBookingStatus.REIMBURSED.value,
+                    ]
+                ),
+                educational_models.CollectiveOffer.status == offer_mixin.CollectiveOfferStatus.EXPIRED.name,
+            )
+        )
+    if DisplayedStatus.EXPIRED.value in statuses:
+        # Status EXPIRED == event is passed without any reservation or cancelled ones
+
+        status_with_booking_filters.append(
+            and_(
+                offer_id_query.c.status.in_(
+                    [
+                        None,
+                        educational_models.CollectiveBookingStatus.PENDING.value,
+                        educational_models.CollectiveBookingStatus.CANCELLED.value,
+                    ]
+                ),
+                educational_models.CollectiveOffer.status == offer_mixin.CollectiveOfferStatus.EXPIRED.name,
+            )
+        )
+    if DisplayedStatus.ARCHIVED.value in statuses:
+        status_filters.append(educational_models.CollectiveOffer.dateArchived != None)
+    if DisplayedStatus.ACTIVE.value in statuses:
+        status_filters.append(
+            and_(
+                educational_models.CollectiveOffer.isActive == True,
+                educational_models.CollectiveOffer.validation == offer_mixin.OfferValidationStatus.APPROVED,
+            )
+        )
+    if DisplayedStatus.REJECTED.value in statuses:
+        status_filters.append(
+            educational_models.CollectiveOffer.validation == offer_mixin.OfferValidationStatus.REJECTED
+        )
+    if DisplayedStatus.PENDING.value in statuses:
+        status_filters.append(
+            educational_models.CollectiveOffer.validation == offer_mixin.OfferValidationStatus.PENDING
+        )
+
+    if status_with_booking_filters:
+        substmt = query_with_booking.filter(or_(*status_with_booking_filters)).subquery()
+        status_filters.append(educational_models.CollectiveOffer.id.in_(sa.select(substmt.c.id)))
+    # we apply all the filters
+    if status_filters:
+        query = query.filter(or_(*status_filters))
+
+    return query
+
+
+def collective_offer_with_booking(
+    query: BaseQuery,
+) -> typing.Tuple[BaseQuery, BaseQuery]:
```

Comment:
Je pense que faire ces changements de noms aiderait à la compréhension. 

Et une petite docstring serait cool

```suggestion
def add_last_booking_status_to_collective_offer_query(
    collective_offer_query: BaseQuery,
) -> typing.Tuple[BaseQuery, BaseQuery]:
```